### PR TITLE
[Editor] Add a toolbar to selected editors with a button to delete it  (bug 1863763)

### LINKED ISSUE #16022: How to undo and redo the changes done by using text box and pencil editor tools in mobile devices. How can I delete some particular change in mobile devices
Attach (recommended) or Link to PDF file here:

Configuration:
- Web browser and its version:
- Operating system and its version:
- PDF.js version: 3.4.406
- Is a browser extension:

Steps to reproduce the problem:
1. Open pdf in mobile devices(iPhon/android)
2. Add text by using Textbox and pencil drawer tool

What is the expected behavior? (add screenshot)
I need to delete pencil drawer changes and text box changes in mobile devices, when pdf is opened.
What went wrong? (add screenshot)
I am not able to delete pencil drawer changes in pdf .
Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

**Comment by calixteman** [comment]
For the moment the aria-label and maybe the tooltip are missing but the strings should be come later.

File: web/annotation_editor_layer_builder.css [hunk 1]
```diff
@@ -95,9 +95,11 @@
 [data-editor-rotation="90"] {
   transform: rotate(90deg);
 }
+
```
Comment (Snuffleupagus) [inline web/annotation_editor_layer_builder.css:98]: Nit: Here, and elsewhere below, there's "unrelated" new-lines inserted.

File: web/annotation_editor_layer_builder.css [hunk 2]
```diff
@@ -188,6 +190,99 @@
...
+    border-radius: 4px;
+    background-color: var(--editor-toolbar-bg-color);
+    border: 1px solid var(--editor-toolbar-border-color);
+    box-shadow: var(--editor-toolbar-shadow);
+
+    &.hidden {
+      display: none;
+    }
+
+    & .icons {
```
Comment (Snuffleupagus) [inline web/annotation_editor_layer_builder.css:244]: In cases such as these, when there's a space between the ampersand and the rule, is the `&` actually necessary?

Looking at https://developer.mozilla.org/en-US/docs/Web/CSS/Nesting_selector#_nesting_selector_and_whitespace it seems that these cases could be written slightly shorter as follows instead, and still achieve the same result?
```suggestion
    .icons {
```

File: web/images/editor-toolbar-bin.svg
Comment (Snuffleupagus) [inline web/images/editor-toolbar-bin.svg:1]: This needs to be added to the gulpfile:
https://github.com/mozilla/pdf.js/blob/42f3d57365dc173d13e54d3e54f6bf8db292adcd/gulpfile.mjs#L1068-L1072

**Review by Snuffleupagus — COMMENTED** [review]
*Leaving a few initial, very quick, observations.*

Naming things is hard, but I feel like calling this `bin` throughout the code is confusing since that mostly describes the *content* of the icon (which may change at some point).
It would rather seems much more appropriate to describe *the functionality*, and thus call this e.g. `delete` (or `remove`) throughout all the code and in the filename.

File: web/annotation_editor_layer_builder.css (hunk 2 above)
Comment (calixteman) [reply web/annotation_editor_layer_builder.css:244]: Cool, thank you for the tip.
Maybe the css linter could remove these useless `&` for us.

File: src/display/editor/toolbar.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ */
+
+import { noContextMenu } from "../display_utils.js";
+
+class EditorToolbar {
+  #div = null;
+
+  #editor;
+
+  #icons = null;
```
Comment (Snuffleupagus) [inline src/display/editor/toolbar.js:23]: It feels slightly strange to, in the `render`-method below, append *buttons* to an element called "icons".
Should we name this property e.g. `#buttons` (or possibly `#container`) instead, and similar for the CSS rules?

File: web/annotation_editor_layer_builder.css [hunk 4]
```diff
@@ -188,6 +188,101 @@
...
+        &:active {
+          border-radius: 2px;
+          background-color: var(--editor-toolbar-active-bg-color);
+        }
+
+        &:focus-visible {
+          border-radius: 3px;
+          outline: 2px solid var(--editor-toolbar-focus-outline-color);
+        }
+      }
```
Comment (Snuffleupagus) [inline web/annotation_editor_layer_builder.css:283]: Can't this block be combined with the one above?

File: web/annotation_editor_layer_builder.css (hunk 2 above)
Comment (Snuffleupagus) [reply web/annotation_editor_layer_builder.css]: > Maybe the css linter could remove these useless `&` for us.

Unfortunately there doesn't appear to be any special rules in Stylelint for the CSS nesting feature, I'm guessing it's still too new.

File: web/annotation_editor_layer_builder.css (hunk 4 above)
Comment (calixteman) [reply web/annotation_editor_layer_builder.css:283]: My idea was to prepare the future, but let me slightly change that stuff.

**Comment by Snuffleupagus** [comment]
After playing with this a bit, I've got two observations/questions:

 - If you select *multiple* editors and then click on the new delete-button only that specific editor is deleted, which while reasonable behaviour in itself *may* possibly be confusing to users.
   Thus I wonder if we need some way to visually notify the user that only *one* editor will be removed, or if this is implicitly understood because there's multiple delete-buttons visible in this case (i.e. one for each selected editor)?

   (Compare with the situation where you select multiple editors and then press the <kbd>Delete</kbd> key, where *all selected* editors are being removed.)

 - *I've not looked in detail at the design-spec, so it's possible that you've simply implemented what the specification says.*
 
   Would it make sense to give the editToolbar-element itself *a little bit* of padding (e.g. `1px` all around), since in the `focus-visible`-state the button-outline and the toolbar-border kind of "blends" together now?
   <br/>
      blend

**Comment by calixteman** [comment]
> After playing with this a bit, I've got two observations/questions:
> 
>     * If you select _multiple_ editors and then click on the new delete-button only that specific editor is deleted, which while reasonable behaviour in itself _may_ possibly be confusing to users.
>       Thus I wonder if we need some way to visually notify the user that only _one_ editor will be removed, or if this is implicitly understood because there's multiple delete-buttons visible in this case (i.e. one for each selected editor)?
>       (Compare with the situation where you select multiple editors and then press the Delete key, where _all selected_ editors are being removed.)
> 

If the user has selected several editors and then click on the delete button I think it makes sense to delete all of them.

>     * _I've not looked in detail at the design-spec, so it's possible that you've simply implemented what the specification says._
>       Would it make sense to give the editToolbar-element itself _a little bit_ of padding (e.g. `1px` all around), since in the `focus-visible`-state the button-outline and the toolbar-border kind of "blends" together now?
>       

Yes, the border line has been recently added to the specs because it helps to have a contrast between the toolbar and a clear background.
The css `:has` has landed few weeks ago in nightly, so it's time to use it in removing the border when the a button has a focus-outline.

File: test/integration/freetext_editor_spec.mjs [hunk 5]
```diff
@@ -3053,4 +3053,138 @@ describe("FreeText Editor", () => {
       );
     });
   });
+
+  describe("Delete a freetext in using the bin button", () => {
```
Comment (Snuffleupagus) [inline test/integration/freetext_editor_spec.mjs:3057]:
```suggestion
  describe("Delete a freetext in using the delete button", () => {
```

File: test/integration/freetext_editor_spec.mjs [hunk 6]
```diff
@@ -3053,4 +3053,138 @@ describe("FreeText Editor", () => {
...
+
+          await page.waitForSelector(getEditorSelector(0), {
+            visible: true,
+          });
+        })
+      );
+    });
+  });
+
+  describe("Delete two freetexts in using the bin button and the keyboard", () => {
```
Comment (Snuffleupagus) [inline test/integration/freetext_editor_spec.mjs:3114]:
```suggestion
  describe("Delete two freetexts in using the delete button and the keyboard", () => {
```

File: web/annotation_editor_layer_builder.css [hunk 7]
```diff
@@ -188,6 +188,105 @@
...
+          background-color: var(--editor-toolbar-hover-bg-color);
+        }
+
+        &:active {
+          border-radius: 2px;
+          background-color: var(--editor-toolbar-active-bg-color);
+        }
+
+        &:focus-visible {
+          border-radius: 3px;
```
Comment (Snuffleupagus) [inline web/annotation_editor_layer_builder.css:284]: With the `:has` rule now added in the latest version, do we still need a different border-radius in this case or could we just specify `border-radius: 2px;` unconditionally at the `.delete` level and remove some duplication?

Comment (calixteman) [reply web/annotation_editor_layer_builder.css:284]: It's a weird thing but border-radius is used for outlines either:
https://developer.mozilla.org/en-US/docs/Web/CSS/outline#description
So I think we must keep it like this.

**Review by Snuffleupagus — APPROVED** [review]
> For the moment the aria-label and maybe the tooltip are missing but the strings should be come later.

Should we wait for the l10n-strings to be provided first, or did you intend for this PR to land as-is (with a later follow-up patch)?

---

r=me, since this seems to work as it should; thank you!

**Comment by calixteman** [comment]
I hope we'll have the strings next week but it's worth having asap in nightly in order to be tested.
So I'll do a follow-up patch once I've them.

**Comment by adityadees** [comment]
@Snuffleupagus If we do a  pencil bruhs mutiple times, I mean like I wrote the checkmark,  it will auto block all, any way to unmerge it?
image

**Pull request merged**